### PR TITLE
surgical linker: don't generate dynhost if it's unchanged

### LINKED ISSUE #3883: Stop surgical linker from breaking caching
Currently the surgical linker will break Rust caching. The issue is that the rust app compiles against a dylib that is generated by the surgical linker. Since we regenerate the the dylib on every run, the rust app final linking step will run every time even if nothing truly changed. As such, without the `precompiled-host` flag, we will waste a lot of timing doing unnecessary work.

We do need to regenerate the dylib on occassion when the platform api changes, but otherwise, it should remain static. Not sure the best way to fix this, but some sort of check to avoid generating the same dylib again may be necessary.

Note: this is just my current understanding of the issue, more investigation may lead to a different root cause

### PR DESCRIPTION
**Pull request by folkertdev**
This is an attempt to fix #3883, though I'm not totally sure it does. At least for me, I don't see faster rebuilds of rust hosts, but also the rebuild time is already very close to zig.

So, I'm not sure it fixes that problem, but I think it's a useful feature anyway.

### CODE REVIEW

File: crates/linker/src/lib.rs [hunk 1]
```diff
@@ -136,10 +136,62 @@ fn generate_dynamic_lib(
...
+            let expected = match target.operating_system {
+                TLO::Linux => object::BinaryFormat::Elf,
+                other => internal_error!("unexpected target {:?}", other),
+            };
+
+            actual == expected
+        }
+        TLA::Aarch64(_) => object.architecture() == object::Architecture::Aarch64,
+        _ => false,
+    }
```
Comment (folkertdev) [inline crates/linker/src/lib.rs:167]: the fragile bit of this code

File: crates/linker/src/lib.rs [hunk 2]
```diff
@@ -136,10 +136,62 @@ fn generate_dynamic_lib(
...
+
+    // the user may have been cross-compiling.
+    // The dynhost on disk must match our current target
+    if !object_matches_target(target, &object) {
+        return false;
+    }
+
+    // we made this dynhost file. For the file to be the same as what we'd generate,
+    // we need all symbols to be there and in the correct order
+    let dynamic_symbols: Vec<_> = object.exports().unwrap();
```
Comment (folkertdev) [inline crates/linker/src/lib.rs:189]: this is quite convenient if you just want the names of exported symbols (sadly it does not give actual `Symbol`s)

**Comment by bhansconnect** [comment]
To fix https://github.com/roc-lang/roc/issues/3883, I think we need to avoid generating the dummy lib, not the dynhost. They dummy lib is a dependency in `build.rs` for rust hosts. So if it changes it causes the host to rebuild.

File: crates/linker/src/lib.rs (hunk 1 above)
Comment (bhansconnect) [reply crates/linker/src/lib.rs:167]: can we put a todo! or unimplemented! instead?

File: crates/linker/src/lib.rs [hunk 3]
```diff
@@ -136,10 +136,62 @@ fn generate_dynamic_lib(
...
+            };
+
+            actual == expected
+        }
+        TLA::Aarch64(_) => object.architecture() == object::Architecture::Aarch64,
+        _ => false,
+    }
+}
+
+fn dynhost_is_up_to_date(target: &Triple, dummy_lib_path: &Path, custom_names: &[String]) -> bool {
```
Comment (bhansconnect) [inline crates/linker/src/lib.rs:170]: Does this say something about dynhost, or just the dummy lib? Just a bit confused by the naming?

**Review by bhansconnect — APPROVED** [review]
If this doesn't work, but you are able to confirm that we avoid updating the dummy lib, my next guess would be that something we do manually when we pass `Some(&dummy_lib)` into `rebuild_host` is messing up the rust caching.

Anyway, my comments are nit.

File: crates/linker/src/lib.rs (hunk 3 above)
Comment (folkertdev) [reply crates/linker/src/lib.rs]: yes sorry mixed those up. made it clearer now

File: crates/linker/src/lib.rs (hunk 1 above)
Comment (folkertdev) [reply crates/linker/src/lib.rs]: yes

**Comment by folkertdev** [comment]
I updated the PR, but as I commented on the original issue, I'm not really sure this is a cargo issue. What sort of setup did you use to see how long cargo spends rebuilding the platform? 

I believe something on the order of 100ms is normal, even if cargo does not actually do anything (using strace on such a cargo invocation is informative: it checks a lot of timestamps etc)

**Comment by bhansconnect** [comment]
Ok. This turned out to be a two part issue. I have been testing with rbt on linux, and it spends about 2 seconds rebuilding the host every run. This is time actually spent in cargo and not in the surgical linker. This obviously was a caching issue.

I thought that a build script that just specified it depended on a dynamic library would only rerun if the dynamic library changes. That is no the case. A rust build script by default will rerun if any file in a package changes. Since the surgical linker dumps files into the platform directory, this meant that the build script would always rerun.

To fix this, we need your change here, but also, we need to scope rust rebuilds. So this change plus adding `println!("cargo:rerun-if-changed=libapp.so")` to `build.rs` fixes the issue for me.

**Review by bhansconnect — APPROVED** [review]

**Comment by folkertdev** [comment]
right, build.rs does not have good caching by default

**Pull request merged**